### PR TITLE
refactor(compiler): support local compilation for @defer blocks

### PR DESCRIPTION
**Pull request by AndrewKushnir**
This commit updates Angular compiler to support local compilation mode for `@defer` blocks.

In local compilation mode, compiler can only rely on the information present in a currently processed file. As a result, compiler can not detect which components/directives/pipes belong to a particular `@defer` block (because that requires selector matching and selector info from other files where those classes are defined) and thus which classes can be defer-loaded. Because of this, no dependency functions are being generated for `@defer` blocks in the local compilation mode.

Also, this commit brings extra configuration and logic to make `@defer` blocks compatible with more constrained local compilation mode internally.

Overall, this change does *not* affect the behavior and APIs of `@defer` blocks externally (outside of Google's codebase) when using regular (full) compilation mode.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts [hunk 1]
```diff
@@ -52,6 +52,8 @@ type UsedNgModule = R3NgModuleDependencyMetadata&{
 
 type AnyUsedType = UsedPipe|UsedDirective|UsedNgModule;
 
+type ComponentTemplate = R3ComponentMetadata<any>['template'];
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:56]: Prefer passing a real type here as opposed to `any` (e.g. `R3TemplateDependency`).

File: packages/compiler/src/render3/view/t2_binder.ts [hunk 2]
```diff
@@ -68,6 +68,11 @@ class Scope implements Visitor {
    */
   readonly namedEntities = new Map<string, Reference|Variable>();
 
+  /**
+   * Set of elements that belong to this scope.
+   */
+  readonly elementsInScope = new WeakSet<Element>();
```
Comment (alxhub) [inline packages/compiler/src/render3/view/t2_binder.ts:74]: Why `WeakSet`?

File: packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts [hunk 3]
```diff
@@ -294,7 +295,10 @@ export class ComponentDecoratorHandler implements
     }
 
     let resolvedImports: Reference<ClassDeclaration>[]|null = null;
+    let resolvedDeferredImports: Reference<ClassDeclaration>[]|null = null;
+
     let rawImports: ts.Expression|null = component.get('imports') ?? null;
+    let rawDeferredImports: ts.Expression|null = component.get('deferredImports') ?? null;
 
     if (rawImports && !metadata.isStandalone) {
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:303]: We should do a similar check for `deferredImports` (assuming it's passed and `imports` isn't).

**Review by alxhub — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:595450614:BASE:597100551:1704853692373:ad79d5a1).

**Review by jessicajaniuk — APPROVED** [review]
reviewed-for: public-api

**Comment by AndrewKushnir** [comment]
**Caretaker notes**:

- [TGP is "green"](https://fusion2.corp.google.com/presubmit/595450614/OCL:595450614:BASE:597121500:1704861442586:bd600f75/targets)
- Alex's and Jess' approval should be sufficient
- this PR is ready for merge

**Comment by atscott** [comment]
This PR was merged into the repository by commit 79fbc4eaba729f7edd3577d6aec22f8e621cf594.

**Pull request closed without merging**